File: packages/sdk-python/src/openllm/_baggage.py
"""OpenTelemetry baggage helpers for multi-tenant context propagation.

Every LLM call in OpenLLM Metrics carries ``tenant``, ``team``, ``app``, ``env``,
and ``project`` so downstream pipelines (gateway, scoring worker, policy
evaluator, audit log) can attribute usage without re-deriving it. Baggage is
the OTel-native way to thread that context through async boundaries and across
process hops.
"""

from __future__ import annotations

from typing import Iterable

from opentelemetry import baggage, context as otel_context
from opentelemetry.context import Context

from ._semconv import (
    BAGGAGE_APP,
    BAGGAGE_ENV,
    BAGGAGE_PROJECT,
    BAGGAGE_TEAM,
    BAGGAGE_TENANT,
)

_TENANT_KEYS: tuple[tuple[str, str], ...] = (
    ("tenant", BAGGAGE_TENANT),
    ("team", BAGGAGE_TEAM),
    ("app", BAGGAGE_APP),
    ("env", BAGGAGE_ENV),
    ("project", BAGGAGE_PROJECT),
)


def attach_tenant_baggage(
    *,
    tenant: str | None,
    team: str | None,
    app: str | None,
    env: str | None,
    project: str | None,
    parent: Context | None = None,
) -> object:
    """Attach tenant context to OTel baggage and return a detach token.

    Empty or ``None`` values are skipped so a missing dimension does not
    overwrite a value already set higher in the call stack.

    The returned token MUST be passed to :func:`detach_tenant_baggage` to
    avoid leaking the baggage into unrelated work.
    """
    ctx = parent if parent is not None else otel_context.get_current()
    values: Iterable[tuple[str, str | None]] = (
        (BAGGAGE_TENANT, tenant),
        (BAGGAGE_TEAM, team),
        (BAGGAGE_APP, app),
        (BAGGAGE_ENV, env),
        (BAGGAGE_PROJECT, project),
    )
    for key, value in values:
        if value:
            ctx = baggage.set_baggage(key, value, context=ctx)
    return otel_context.attach(ctx)


def detach_tenant_baggage(token: object) -> None:
    """Detach a baggage scope previously attached by :func:`attach_tenant_baggage`."""
    otel_context.detach(token)  # type: ignore[arg-type]


def current_tenant_baggage() -> dict[str, str]:
    """Return the tenant-context keys currently set on OTel baggage.

    Keys are the short-form names (``tenant``, ``team``, ``app``, ``env``,
    ``project``); values are the strings stored on baggage. Missing keys are
    omitted from the returned dict.
    """
    result: dict[str, str] = {}
    for short, full in _TENANT_KEYS:
        value = baggage.get_baggage(full)
        if isinstance(value, str) and value:
            result[short] = value
    return result

File: packages/sdk-go/baggage.go
// Copyright 2026 Yasvanth Udayakumar
// Licensed under the Apache License, Version 2.0.

package openllm

import (
	"context"

	"go.opentelemetry.io/otel/baggage"
)

// Baggage keys for multi-tenant context propagation. Downstream services pick
// these up from the wire to attribute usage without re-deriving it. The names
// match the .NET, Python, and Node.js SDKs.
const (
	BaggageTenant  = "openllm.tenant"
	BaggageTeam    = "openllm.team"
	BaggageApp     = "openllm.app"
	BaggageEnv     = "openllm.env"
	BaggageProject = "openllm.project"
)

// attachTenantBaggage returns a child context carrying tenant/team/app/env/
// project on OTel baggage. Empty values are skipped so a missing dimension at
// this layer never overwrites one set higher in the call stack. Any baggage
// build error is logged via the OTel error handler and the original context is
// returned — instrumentation must never break the caller.
func attachTenantBaggage(ctx context.Context, c CallOptions) context.Context {
	bag := baggage.FromContext(ctx)
	bag = setMemberIfPresent(bag, BaggageTenant, c.Tenant)
	bag = setMemberIfPresent(bag, BaggageTeam, c.Team)
	bag = setMemberIfPresent(bag, BaggageApp, c.App)
	bag = setMemberIfPresent(bag, BaggageEnv, c.Env)
	bag = setMemberIfPresent(bag, BaggageProject, c.Project)
	return baggage.ContextWithBaggage(ctx, bag)
}

// setMemberIfPresent adds a baggage member if value is non-empty, returning the
// existing baggage unchanged on either an empty value or a build error.
func setMemberIfPresent(b baggage.Baggage, key, value string) baggage.Baggage {
	if value == "" {
		return b
	}
	member, err := baggage.NewMember(key, value)
	if err != nil {
		return b
	}
	out, err := b.SetMember(member)
	if err != nil {
		return b
	}
	return out
}

// CurrentTenantBaggage returns the tenant-context keys currently set on OTel
// baggage in ctx. Keys are short-form names (tenant/team/app/env/project);
// missing keys are omitted.
func CurrentTenantBaggage(ctx context.Context) map[string]string {
	b := baggage.FromContext(ctx)
	out := make(map[string]string, 5)
	for short, full := range map[string]string{
		"tenant":  BaggageTenant,
		"team":    BaggageTeam,
		"app":     BaggageApp,
		"env":     BaggageEnv,
		"project": BaggageProject,
	} {
		if m := b.Member(full); m.Value() != "" {
			out[short] = m.Value()
		}
	}
	return out
}
